### PR TITLE
Read win-title in non-blocking manner

### PR DESCRIPTION
**Pull request by N-R-K**
* rather than calling the script unconditionally per redraw, we now have
  a `title_dirty` flag and keep track of when any of the relavent
  information changes.

* instead of reading from `win-title` directly and potentially getting
  blocked, we poll on it similar to `{image,thumb}-info` and only call
  read on it when we're sure there's data to be read.

to reproduce the 2nd issue, add `sleep 1` at the start of `win-title`
script and notice how it ends up freezing the UI.

Co-authored-by: Arthur Williams
Closes: https://github.com/nsxiv/nsxiv/issues/258

### CODE REVIEW

File: main.c [hunk 1]
```diff
@@ -75,17 +75,15 @@ static struct {
 	int fd;
 	unsigned int i, lastsep;
 	pid_t pid;
-} info;
+} info, wintitle;
```
Comment (N-R-K) [inline main.c:78]: The `i, lastsep` members aren't used for `wintitle` but they're getting removed in #271 anyways. So I thought I'd merge `info` and `wintitle` together.

**Comment by TAAPArthur** [comment]
>instead of reading from win-title directly and potentially getting blocked, we poll on it similar to {image,thumb}-info

I think this is a questionable change. Wanting the script to block is defendable behavior. If I'm remembering correctly, we add the win-title ourselves so we __could__ change it, but think that should be a separate commit/discussion.

Some concerns I have:

- What if you actually don't want nsix to proceed until the command finishes? If your command was really slow your window title could be out of date. 
- If we call the win-title script, it takes a while, nsxiv continues and triggers a redraw before the first script finishes, what should happen? Do we have 2 running? Do we not start the 2nd? Do we kill the first and then start the 2nd
- Does anyone actually care about whether the script blocks in practice? If not then it seems simpler to just wait for it to complete.

Please note that these scenarios are devoid of any examples. I can't really come up with a case where the script would really have to take a long time. And if such a case where did happen, I'm pretty sure a user could solve it themselves. (Ie background the script, cache the result and/or manually update the nsxiv title).

So given the dirty flag solves the presented issue sufficiently well, I'd say we just do that and comeback to this other issue if we care. I'd argue to go the other way and make all the scripts block just cause it makes everything simpler

**Comment by TAAPArthur** [comment]
Think you're last 2 pros are a matter of opinion. 

>And even after considering the arguments above if this is still seen as controversial then I'm not opposed creating a new PR separating the dirty and the non-blocking part.

Thanks

Regardless of it going it or not, I think it should be a separate commit/PR simply because the dirty fix is sufficient to solve the linked issue. But in regards to the question of if we should block, do we expect people to actually have heavy infos scripts? If not we can simply the code for no real penalty. And if so we probably should document our behavior in the case an excessively long command is spawned. So it's less about win-title should block or not but weather if image/thumb-info should also. I do agree that they probably should all have the same behavior

**Comment by N-R-K** [comment]
> So it's less about win-title should block or not but weather if image/thumb-info should also.

Non-blocking is simpler, sure. But I suppose we shouldn't change image-info since doing so will introduce regression from `sxiv` for anyone who's doing some heavy processing in there.

But since `win-title` is a nsxiv addition, we can be more liberal with it. But I'm not really convinced the reduced LoC outweighs being consistent with `image-info` + reduction of unnecessary blocking.

Anyhow reopened #266 with the call reduction part.

**Comment by TAAPArthur** [comment]
So we agree on what's simplier and that all the scripts should be the same.

> But I suppose we shouldn't change image-info since doing so will introduce regression from sxiv for anyone who's doing some heavy processing in there.

This seems to be point of debate. Is the current behavior a bug or a feature or an implementation detail.

I think we agree that if it isn't a feature, we should do the simpler.

So supposing that the current behavior is actually useful to someone, I'd then argue that whatever their usecase is could still be done if we blocked. By not blocking, if we try to spawn a script and a prior one hasn't finished:

1. Kill the 1st and run the 2nd
2. Not start the 2nd because the 1st is still ongoing
3. Spawn the 2nd while the 1st is running and have them race to update the title
4. Spawn the 2nd but stop listening for the 1st
5. Wait for the first to complete and then run the second

In any of the above, users could achieve the same result if we blocked with some trivial modifications to their shell scripts (assuming standard POSIX/wm tools. If in doubt I could provide examples.

Going to go out on a limb here and say that options 3/4 both suck since if the script was resource intensive, spawning more of them isn't the answer. 5 isn't any better than just always blocking. 1 is likely to cause process leaks and/or the child could ignore the kill request. And 2 is the easiest to solve in a shell scripts (textbook case of flock)

@N-R-K  And if we disagree here, could you share what the current/ideal behavior would be in a script was `sleep 100; echo $IMAGE_NAME` and I move 3 images over and wait 2 min?

**Comment by explosion-mental** [comment]
> Believe that will be fixed in https://github.com/nsxiv/nsxiv/pull/266

I C, nice

**Comment by N-R-K** [comment]
> So supposing that the current behavior is actually useful to someone, I'd then argue that whatever their usecase is could still be done if we blocked.

I'd avoid breaking things which worked on `sxiv`, so I'd vote no for changing current `image-info` behavior.

> could you share what the current/ideal behavior

The current behavior (with `image-info`) is to kill the first and spawn the 2nd. If we do it with `win-title` it's going to be simpler than that because we won't have `{open,close}_info()` calls scattered around, there's only one place where we update the `win-title`.

**Pull request closed without merging**